enterprise/server/remote_execution/commandutil/commandutil_windows.go

```diff
@@ -0,0 +1,89 @@
+//go:build windows
+
+package commandutil
+
+import (
+	"fmt"
+	"os/exec"
+	"syscall"
+	"unsafe"
+
+	"golang.org/x/sys/windows"
+)
+
+type process struct {
+	cmd        *exec.Cmd
+	terminated chan struct{}
+
+	jobHandle windows.Handle
+}
+
+func (p *process) preStart() error {
+	j, err := createWindowsJobObject()
+	if err != nil {
+		return err
+	}
+
+	p.jobHandle = j
+	return nil
+}
+
+func (p *process) postStart() error {
+	processHandle, err := windows.OpenProcess(windows.PROCESS_QUERY_LIMITED_INFORMATION|windows.PROCESS_TERMINATE, false, uint32(p.cmd.Process.Pid))
+	if err != nil {
+		return nil
+	}
+	defer windows.CloseHandle(processHandle)
+
+	return windows.AssignProcessToJobObject(p.jobHandle, processHandle)
+}
+
+func createWindowsJobObject() (windows.Handle, error) {
+	job, err := windows.CreateJobObject(nil, nil)
+	if err != nil {
+		return windows.InvalidHandle, fmt.Errorf("failed to create job object: %w", err)
+	}
+	limitInfo := windows.JOBOBJECT_BASIC_LIMIT_INFORMATION{
+		LimitFlags: windows.JOB_OBJECT_LIMIT_KILL_ON_JOB_CLOSE,
+	}
+	if _, err := windows.SetInformationJobObject(
+		job,
+		windows.JobObjectBasicLimitInformation,
+		uintptr(unsafe.Pointer(&limitInfo)),
+		uint32(unsafe.Sizeof(limitInfo))); err != nil {
+		return windows.InvalidHandle, fmt.Errorf("failed to set job object info: %w", err)
+	}
+
+	return job, nil
+}
+
+// killProcessTree kills the given pid as well as any descendant processes.
+//
+// It tries to kill as many processes in the tree as possible. If it encounters
+// an error along the way, it proceeds to kill subsequent pids in the tree. It
+// returns the last error encountered, if any.
+//
+// For Windows, see
+// https://learn.microsoft.com/en-us/windows/win32/procthread/job-objects
+// and https://learn.microsoft.com/en-us/windows/win32/procthread/nested-jobs
+// for more details.
+func (p *process) killProcessTree() error {
+	return windows.CloseHandle(p.jobHandle)
+}
```

Comment:
wow, this part is much simpler than unix :)

does it work even if subprocesses "fork" into their own jobs? I think that's the reason I added the "KillProcessTree" thing to begin with - on unix, processes could fork into their own process group, and then killing the process group would not properly clean up all processes.

Reply:
> Operations performed on a job object affect all processes associated with the job object. Examples include enforcing limits such as working set size and process priority or terminating all processes associated with a job.

From https://learn.microsoft.com/en-us/windows/win32/procthread/job-objects

It's also used by other CI vendors such as https://github.com/buildkite/agent/blob/8faa01f13fd92d8a730f1c04bf457ead05c173b1/process/signal_windows.go#L76-L79

I did not set up a specific test for it, mostly because I don't wanna dive too deep into PowerShell arcane magic. But I did put some basic tests in place for us to replicate in the future if anything needs to be reproduced.